### PR TITLE
feat: git material include exclude feature and partial commit search

### PR DESCRIPTION
**Pull request by vikramdevtron**
# Description

While configuration git repository, users may provide you the include and exclude specific paths for builds, allowing only those materials for builds which are within these specific paths.
Filter pattern only applicable on branch fixed source type.

Fixes https://github.com/devtron-labs/devtron/issues/3129

## How Has This Been Tested?
- [x] Fetch Ci materials will fetch after applying Filters
- [x] Fetch Ci materials with fetch after applying Filters, also include excluded items, with marked. 
- [x] Auto trigger will block for excluded materials
- [x] commit search can be found based on partial search also.

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [x] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

**Review by iamayushm — DISMISSED** [review]

**Review by iamayushm — DISMISSED** [review]

File: pkg/pipeline/CiHandler.go [hunk 1]
```diff
@@ -284,7 +285,85 @@ func (impl *CiHandlerImpl) RefreshMaterialByCiPipelineMaterialId(gitMaterialId i
 	return refreshRes, err
 }
 
-func (impl *CiHandlerImpl) FetchMaterialsByPipelineId(pipelineId int) ([]CiPipelineMaterialResponse, error) {
+func (impl *CiHandlerImpl) FetchMaterialsByPipelineIdAndGitMaterialId(pipelineId int, gitMaterialId int, showAll bool) ([]CiPipelineMaterialResponse, error) {
+	ciMaterials, err := impl.ciPipelineMaterialRepository.GetByPipelineIdAndGitMaterialId(pipelineId, gitMaterialId)
+	impl.Logger.Infow("Testing Fetch Ci materials by pipeline Id ", "ci materials", ciMaterials)
```
Comment (kripanshdevtron) [inline pkg/pipeline/CiHandler.go:290]: we can log metadata like pipelineId and git material Id only, why do we need to log materials ??

File: pkg/pipeline/CiHandler.go [hunk 2]
```diff
@@ -284,7 +285,85 @@ func (impl *CiHandlerImpl) RefreshMaterialByCiPipelineMaterialId(gitMaterialId i
...
+	if err != nil {
+		impl.Logger.Errorw("ciMaterials fetch failed", "err", err)
```
Comment (kripanshdevtron) [inline pkg/pipeline/CiHandler.go:292]: we need to log metadata as well

File: api/restHandler/app/BuildPipelineRestHandler.go [hunk 3]
```diff
@@ -1191,6 +1257,17 @@ func (handler PipelineConfigRestHandlerImpl) FetchChanges(w http.ResponseWriter,
 		common.WriteJsonResp(w, err, nil, http.StatusBadRequest)
 		return
 	}
+	showAll := false
+	v := r.URL.Query()
```
Comment (kripanshdevtron) [inline api/restHandler/app/BuildPipelineRestHandler.go:1261]: variable names could have been better

File: pkg/pipeline/CiHandler.go [hunk 4]
```diff
@@ -284,7 +285,85 @@ func (impl *CiHandlerImpl) RefreshMaterialByCiPipelineMaterialId(gitMaterialId i
...
+		if !exists {
+			ciPipelineMaterialResponses = append(ciPipelineMaterialResponses, r)
+		}
+	}
+	return ciPipelineMaterialResponses, nil
+}
+
+func (impl *CiHandlerImpl) FetchMaterialsByPipelineId(pipelineId int, showAll bool) ([]CiPipelineMaterialResponse, error) {
 	ciMaterials, err := impl.ciPipelineMaterialRepository.GetByPipelineId(pipelineId)
 	impl.Logger.Infow("Testing Fetch Ci materials by pipeline Id ", "ci materials", ciMaterials)
```
[the remark this answers is not in the document]
Comment (vikramdevtron) [reply pkg/pipeline/CiHandler.go:368]: log not required here, done

File: pkg/pipeline/CiHandler.go (hunk 2 above)
Comment (vikramdevtron) [reply pkg/pipeline/CiHandler.go:292]: log not required here

**Review by iamayushm — DISMISSED** [review]

**Review by iamayushm — APPROVED** [review]

**Review by kripanshdevtron — APPROVED** [review]

**Pull request merged**